### PR TITLE
fix(QueryContext): validation does not validate query_context metrics

### PR DESCRIPTION
**Pull request by ofekisr**
### SUMMARY

close 
continue in #19753

### CODE REVIEW

**Review by amitmiran137 — CHANGES_REQUESTED** [review]
Extract until you can't

**Review by amitmiran137 — APPROVED** [review]
LGTM !
Feature introduced here will enforce datasource access on custom SQL expression of metrics 

@villebro we should add column in a follow up PR

**Review by craig-rueda — CHANGES_REQUESTED** [review]
Thanks for this PR - it's bumping up security! I don't see any tests, however. Also, there's a lot of single line methods in the ContextValidator. Can we compress those a bit? Too much delegation makes the code a bit hard to follow along with.

File: superset/charts/api.py [hunk 1]
```diff
@@ -617,7 +619,7 @@ def get_data(self, pk: int) -> Response:
         json_body["result_type"] = request.args.get("type", ChartDataResultType.FULL)
 
         try:
-            command = ChartDataCommand()
+            command = self._create_chart_data_command()
```
[the remark this answers is not in the document]
Comment (jayakrishnankk) [reply superset/charts/api.py:622]: What's left TO DO in this, so that it can get merged ?

**Comment by ofekisr** [comment]
@craig-rueda @hughhhh @jayakrishnankk 
After all the recent refactor in ChartData, I recommitted the PR with all the adjustments to refactored code

File: superset/charts/api.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ofekisr) [reply superset/charts/api.py]: @jayakrishnankk review and merge

File: superset/security/manager.py [hunk 2]
```diff
@@ -1047,17 +1048,23 @@ def raise_for_access(
...
-                    feature_flag_manager.is_feature_enabled("DASHBOARD_RBAC")
-                    and self.can_access_based_on_dashboard(datasource)
-                )
-            ):
-                raise SupersetSecurityException(
-                    self.get_datasource_access_error_object(datasource)
-                )
+            self.raise_when_there_is_no_access_to(datasource)
+
+    def raise_when_there_is_no_access_to(self, datasource: BaseDatasource) -> None:
```
Comment (villebro) [inline superset/security/manager.py:1053]: This name should IMO follow the existing naming conventions. In this case `raise_for_datasource_access` would be more appropriate and more aligned with the pre-existing `raise_for_dashboard_access`.

File: tests/common/query_context_generator.py [hunk 3]
```diff
@@ -248,12 +302,77 @@ def generate(
...
+                    "type_generic": 1,
+                    "verbose_name": None,
+                    "warning_markdown": None,
+                },
+                "expressionType": "SQL",
+                "hasCustomLabel": False,
+                "isNew": False,
+                "label": "COUNT({})".format(column_name),
+                "optionName": "metric_bvvzfjgdg7_eawlsdp84tq",
+                "sqlExpression": "(SELECT count(name) FROM {})".format(table_name),
```
Comment (villebro) [inline tests/common/query_context_generator.py:372]: nit: 
```suggestion
                "sqlExpression": f"(SELECT count(name) FROM {table_name})",
```

File: tests/common/query_context_generator.py [hunk 4]
```diff
@@ -248,12 +302,77 @@ def generate(
...
+                    "python_date_format": None,
+                    "type": "TEXT",
+                    "type_generic": 1,
+                    "verbose_name": None,
+                    "warning_markdown": None,
+                },
+                "expressionType": "SQL",
+                "hasCustomLabel": False,
+                "isNew": False,
+                "label": "COUNT({})".format(column_name),
```
Comment (villebro) [inline tests/common/query_context_generator.py:370]: nit:
```suggestion
                "label": f"COUNT({column_name})",
```

File: superset/charts/data/query_context_validator.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    def _validate_actor_can_access(self, query_context: QueryContext) -> None:
+        sql_database: Optional[Database] = query_context.get_database()
+        if sql_database is not None:
+            self._validate_when_context_based_on_sql_database(
+                query_context, sql_database
+            )
+        else:
+            self._security_manager.raise_for_access(query_context=query_context)
+
+    def _validate_when_context_based_on_sql_database(
```
Comment (villebro) [inline superset/charts/data/query_context_validator.py:61]: This has come up in previous review comments from me. While I agree that it's good practice to break out logical chunks to improve readability and enforce separation of concerns, there is a point when abstraction can actually decrease readability. Each nested call causes additional cognitive overhead for the reviewer, and dilutes the relevance of other more important functions/methods. As a rule of thumb, if a private method is 1) only used once 2) consists of less than 5 lines of code 3) further branches out into another private method which also satisfies 1) and 2), I would consider just inlining the code and adding a comment describing what is happening (if needed).

File: superset/security/manager.py [hunk 6]
```diff
@@ -974,14 +976,14 @@ def set_perm(  # pylint: disable=unused-argument
...
-        query: Optional["Query"] = None,
-        query_context: Optional["QueryContext"] = None,
-        table: Optional["Table"] = None,
-        viz: Optional["BaseViz"] = None,
+        database: Optional[Database] = None,
+        datasource: Optional[BaseDatasource] = None,
+        query: Optional[Query] = None,
+        query_context: Optional[QueryContext] = None,
+        table: Optional[Table] = None,
+        viz: Optional[BaseViz] = None,
```
Comment (villebro) [inline superset/security/manager.py:986]: Is this nowadays the correct way of referring to imports from `TYPE_CHECKING`?

**Review by villebro — COMMENTED** [review]
First pass comments. I think this is a great improvement to the security model, but I'm concerned this may cause false positives (I did a similar change some time ago, which caused unpleasant regressions due to triggering false positives in edge cases that hadn't been thought of in the original PR). To make rolling out new security improvements less problematic, I would almost consider introducing (yet another) feature flag "SECURITY_BETA" where new security improvements can be introduced and kept until more comprehensive testing has been done.

File: superset/charts/data/commands/get_data_command.py [hunk 7]
```diff
@@ -14,6 +14,8 @@
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
+from __future__ import annotations
+
```
[the remark this answers is not in the document]
Comment (ofekisr) [reply superset/charts/data/commands/get_data_command.py:18]: I develop with py 3.8, without it raises errors

File: superset/security/manager.py [hunk 8]
```diff
@@ -14,8 +14,10 @@
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
-# pylint: disable=too-many-lines
+# pylint: disable=too-many-lines, invalid-name
```
[the remark this answers is not in the document]
Comment (ofekisr) [reply superset/security/manager.py:17]: I don't think we should use the invalid-name rule as well... 
I use the clean code principles, so there is no sense to put ignoring on each statement

File: superset/security/manager.py (hunk 2 above)
Comment (ofekisr) [reply superset/security/manager.py:1053]: "raise_for_dashboard_acess" is even more bad name. what is mean "dashboard_acess" ?

a better name conventions are on of those: validate_<object_type>_can_be_accessed
or 
actor_can_access_<object_type>

File: superset/security/manager.py (hunk 6 above)
Comment (ofekisr) [reply superset/security/manager.py:986]: a string representation is worst than the real type

Comment (betodealmeida) [reply superset/security/manager.py:986]: If the imports are inside an `if TYPE_CHECKING` then you need to use a string, otherwise you get a `NameError`.

File: superset/security/manager.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (betodealmeida) [reply superset/security/manager.py:17]: Putting on each statement means you're aware of the rule, and you consciously decided to break it because practicality beats purity. If you put it at the top of the model the rule could be broken many times without people being aware of it, which reduces the code quality.

File: superset/security/manager.py (hunk 2 above)
Comment (betodealmeida) [reply superset/security/manager.py:1053]: We can discuss which naming strategy is better, but I think we can all agree that consistency is a good thing. If we have an existing pattern I think it would be better to stick with it, since it makes it easier for people to understand the code.

File: superset/charts/data/commands/get_data_command.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (craig-rueda) [reply superset/charts/data/commands/get_data_command.py:18]: Yep, let's remove this if we can.

File: superset/charts/data/query_context_validator.py (hunk 5 above)
Comment (craig-rueda) [reply superset/charts/data/query_context_validator.py:61]: Yep, I agree. Too much delegation makes the logic flow hard to follow

File: superset/security/manager.py (hunk 2 above)
Comment (etr2460) [reply superset/security/manager.py:1053]: noting that this pattern was originally used (i think) because it follows the similar one used by the python requests package, `raise_for_status`: https://docs.python-requests.org/en/latest/api/#requests.Response.raise_for_status

@john-bodley might know a bit more about why we picked this naming

File: superset/security/manager.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (john-bodley) [reply superset/security/manager.py:17]: +1. The desired pattern is mentioned [here](https://github.com/apache/superset/blob/master/CONTRIBUTING.md#python).

File: superset/charts/data/commands/get_data_command.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ofekisr) [reply superset/charts/data/commands/get_data_command.py:18]: https://docs.python.org/3/library/__future__.html

image

File: superset/security/manager.py (hunk 2 above)
Comment (ofekisr) [reply superset/security/manager.py:1053]: @etr2460 
You gave a good reference, but I still think is an anti-pattern, and they are wrong too.
but with that reference, it should be discussed in a different context not in that PR

@betodealmeida 
There is a conflict between two principles: 
consistency vs you should always leave the code you have just modified cleaner than it was, a good catch and currently, I don't have a good argument or opinion on what principle is stronger. 

I'll rename it

File: superset/security/manager.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ofekisr) [reply superset/security/manager.py:17]: I'll add and open new discussion on it

File: superset/charts/data/commands/get_data_command.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ofekisr) [reply superset/charts/data/commands/get_data_command.py:18]: @villebro 
regarding the FF, I don't agree with it since the change is fixing a bug a not adding new functionality.
if someone will encounter a surprising issue with it, it means "hey it got something, we have a security breach" so the app administrator should think about how to solve it, giving extra permission or removing the metric.

this is my IMO, but for now I'll add that FF

[the remark this answers is not in the document]
Comment (villebro) [reply superset/charts/data/commands/get_data_command.py:18]: @ofekisr My point for introducing the FF is to gradually introduce the new stricter security check in a way which makes it possible to catch potential regressions that weren't caught during initial review. By having an introductory period during which the stricter checks are behind a FF, orgs that require the new stricter checking can help harden the feature by pushing fixes to master before the new checks are made generally available.

File: superset/security/manager.py (hunk 6 above)
Comment (ofekisr) [reply superset/security/manager.py:986]: > 
When you use the old method of type hint as string, you actually don't need to import the class at all and don't need to put it under TYPE_CHECKING
but if you would like to use the new better choice for marking the type hint by the real type 
you should import them under TYPE_CHECKING with the __future__ annotations.
That is the better choice so you can navigate and let the IDEs do their job --> assist you

**Comment by jayakrishnankk** [comment]
@craig-rueda - requesting review. bumping this.

File: tests/common/query_context_generator.py [hunk 9]
```diff
@@ -199,16 +199,16 @@ def get_query_object(
...
-    obj = QUERY_OBJECTS[query_name]
-
-    # apply overrides
-    if ":" in query_name:
-        parent_query_name = query_name.split(":")[0]
-        obj = {
-            **QUERY_OBJECTS[parent_query_name],
-            **obj,
-        }
+        obj = QUERY_OBJECTS[query_name]
```
Comment (villebro) [inline tests/common/query_context_generator.py:202]: I don't understand the point of this change - if `query_name not in QUERY_OBJECTS`, we probably can't assign `obj = QUERY_OBJECTS[query_name]`? Consequently, the code below becomes unreachable, raising the question if it's needed anymore

Comment (ofekisr) [reply tests/common/query_context_generator.py:202]: this is not the updated code
